core/common/scale.cpp

```diff
@@ -0,0 +1,506 @@
+        enum type : size_t {
+          kLessThan64 = 1ul << 6,
+          kLessThan16384 = 1ul << 14,
+          kLessThanTwoDegreeThirty = 1ul << 30
+        };
+      };
+
+      template <class T>
+      struct EncodedLengthCalculator {
+        EncodedLengthCalculator() {}
+
+        size_t operator()(const T &) {
+          assert(false && "not implemented for this type");
+          return 0;
+        }
+      };
+
+      EncodeIntegerResult encodeUint8(uint8_t value) noexcept {
+        if (value < EncodingCategoryLimits::kLessThan64) {
+          const uint8_t v = (value << 2);
+          return expected::Value{ByteArray{v}};
+        }
+
+        return encodeUint16(static_cast<uint16_t>(value));
+      }
+
+      EncodeIntegerResult encodeInt8(int8_t value) noexcept {
+        return encodeUint8(util::makeSignReinterpreter(value).asUnsigned());
+      }
+
+      EncodeIntegerResult encodeUint16(uint16_t value) noexcept {
+        if (value < EncodingCategoryLimits::kLessThan16384) {
+          uint16_t v = value;
+          v <<= 2;  // v *= 4
+          ++v;      // set 0b01 flag
+          auto minor_byte = static_cast<uint8_t>(v & 0xFF);
+          v >>= 8;
+          auto major_byte = static_cast<uint8_t>(v & 0xFF);
+          return expected::Value{ByteArray{minor_byte, major_byte}};
+        }
+
+        return encodeUint32(static_cast<uint32_t>(value));
+      }
+
+      EncodeIntegerResult encodeInt16(int16_t value) noexcept {
+        return encodeUint16(util::makeSignReinterpreter(value).asUnsigned());
+      }
+
+      EncodeIntegerResult encodeUint32(uint32_t value) noexcept {
+        if (value < EncodingCategoryLimits::kLessThanTwoDegreeThirty) {
+          uint32_t value_le = impl::encodeLittleEndian((value << 2) + 2);
+
+          impl::AsArray<uint32_t> arr{value_le};
+
+          return expected::Value{ByteArray(arr.begin(), arr.end())};
+        }
+
+        return encodeUint64(static_cast<uint64_t>(value));
+      }
+
+      EncodeIntegerResult encodeInt32(int32_t value) noexcept {
+        return encodeUint32(util::makeSignReinterpreter(value).asUnsigned());
+      }
+
+      EncodeIntegerResult encodeUint64(uint64_t value) noexcept {
+        return encodeBigInteger(BigInteger{value});
+      }
+
+      EncodeIntegerResult encodeInt64(int64_t value) noexcept {
+        if (value < 0)
+          return expected::Error{EncodeBigIntegerError::kBigNegativeValue};
+
+        return encodeUint64(static_cast<uint64_t>(value));
+      }
+
+      EncodeIntegerResult encodeBigInteger(const BigInteger &value) noexcept {
+        // cannot encode negative numbers
+        if (value < 0)
+          return expected::Error{EncodeBigIntegerError::kBigNegativeValue};
+
+        // calculate number of bytes required
+        auto countBytes = [](BigInteger v) -> size_t {
+          if (v == 0)
+            return 1;
+
+          size_t counter = 0;
+          while (v > 0) {
+            ++counter;
+            v /= 256;
```

Comment:
`v >>= 8` is more intuitive here, since we're counting bytes

Reply:
Ok